### PR TITLE
Prevent drivers from streaming stdout/stderr to other drivers.

### LINKED ISSUE #4029: Stdout/stderr from all drivers are streamed to all other drivers.
As of https://github.com/ray-project/ray/pull/3892, the stdout/stderr of all drivers is streamed to all other drivers. An individual driver can opt out by calling `ray.init(..., log_to_driver=False)`.

To reproduce this, do the following.

1. Start Ray.

    ```
    ray start --head --redis-port=6379
    ```

2. Start a driver that prints stuff.

    ```python
    import ray
    import sys
    import time
 
   ray.init(redis_address='localhost:6379')

    @ray.remote
    def f():
        while True:
            print('hi')
            sys.stdout.flush()
            time.sleep(0.1)

    f.remote()
    ```

3. Start another driver, and see that it receives the stdout from the first driver.

    ```python
    import ray

    ray.init(redis_address='localhost:6379')
    ```

### Some options to address this.

1. Workers can belong to specific drivers (this means logs will be complicated for cross driver applications, e.g., named actors).
2. Workers can redirect there stdout/stderr on a per-task basis.
3. Workers can print extra annotations in their stdout/stderr which the log monitor can parse and use to choose which drivers to publish to.

### PR DESCRIPTION
**Pull request by ajgokhale**
## What do these changes do?

Stdout/stderr from each worker is streamed to the driver that initiated the task assigned to the worker, instead of all drivers in the cluster.

## Related issue number

Closes #4029.

## Linter

- [X] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/log_monitor.py [hunk 1]
```diff
@@ -33,6 +33,7 @@ def __init__(self,
         self.file_position = file_position
         self.file_handle = file_handle
         self.worker_pid = None
+        self.driver_id = None
```
Comment (robertnishihara) [inline python/ray/log_monitor.py]: Let's call this `current_driver_id`. Can you also add a comment on the previous line explaining what the purpose of this field is?

File: python/ray/log_monitor.py [hunk 2]
```diff
@@ -96,11 +97,10 @@ def update_log_filenames(self):
             if full_path not in self.log_filenames:
                 self.log_filenames.add(full_path)
                 self.closed_file_infos.append(
-                    LogFileInfo(
-                        filename=full_path,
-                        size_when_last_opened=0,
-                        file_position=0,
-                        file_handle=None))
+                    LogFileInfo(filename=full_path,
```
Comment (robertnishihara) [inline python/ray/log_monitor.py]: There are a bunch of formatting changes in this PR. I think you probably have a newer version of `yapf` installed. Can you install `yapf==0.23.0` and rerun the formatting? This is the version of yapf we use in Travis.

File: python/ray/worker.py [hunk 3]
```diff
@@ -994,6 +993,11 @@ def _wait_for_and_process_task(self, task):
                 next_title = "ray_{}".format(actor.__class__.__name__)
             with profiling.profile("task", extra_data=extra_data):
                 with _changeproctitle(title, next_title):
+                    print("Ray driver id: {}".format(str(driver_id)))
```
Comment (robertnishihara) [inline python/ray/worker.py]: This is being printed for every task, right? Can we print it only when the driver ID changes? Otherwise the log files will grow in proportion to the number of tasks.

File: python/ray/tests/test_basic.py [hunk 4]
```diff
@@ -2536,26 +2534,53 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+        if i >= j * 100 and i < (j + 1) * 100:
+            assert str(-i) in output_lines
+        else:
+            assert str(-i) not in output_lines
     error_lines = captured["err"]
     assert len(error_lines) == 0
 
 
+def test_logging_to_multiple_drivers(shutdown_only):
+    file_name = os.path.abspath(__file__) + "::test_logging_to_driver"
```
Comment (robertnishihara) [inline python/ray/tests/test_basic.py]: Interesting approach. I have some suggestions for making it more consistent with the other tests.

One minor issue is that `python` may not actually find the right `python` command, e.g., if the user started the test using `python3` or something like that. To handle this kind of thing in other places in the codebase, we use `sys.executable`.

Take a look at how the tests are written in e.g., https://github.com/ray-project/ray/blob/master/python/ray/tests/test_multi_node.py

- This uses the `call_ray_start` pytest fixture to start a Ray cluster in the background.
- Then we can connect the test as a driver to the cluster with `ray.init(redis_address=...)`
- Then we can start several drivers in the background that do some printing (both to stdout and stderr) using `run_string_as_driver` or `run_string_as_driver_nonblocking` (2 or three should be sufficient).
- We can capture all of the main test scripts stdout/stderr using `with CaptureOutputAndError` like in `test_logging_to_driver`.
- We can run some tasks from the main script that do some logging.
- We can make sure that we do see the output from the tasks that the main script submitted. And make sure that we don't see the output from the tasks that the background drivers submitted.

**Review by robertnishihara — COMMENTED** [review]
Hey @ajgokhale, this looks fantastic! I left some comments.

File: python/ray/log_monitor.py [hunk 5]
```diff
@@ -170,15 +170,30 @@ def check_log_files_and_publish_updates(self):
...
+            for idx in range(max_num_lines_to_read):
                 next_line = file_info.file_handle.readline()
                 if next_line == "":
                     break
                 if next_line[-1] == "\n":
                     next_line = next_line[:-1]
+
+                # Record when a worker switches drivers
+                if next_line.startswith("Ray driver id: "):
+                    driver_id = next_line.split(" ")[-1]
```
Comment (robertnishihara) [inline python/ray/log_monitor.py]: One way that things could go wrong (no need to address this right now, I'm not actually sure how we would address that other than having separate workers for different drivers), is if a task starts a background thread that just runs forever and prints stuff in the background. Then that print statement could corrupt the `"Ray driver id: " print statement and then mess up what the log monitor does.

File: python/ray/worker.py (hunk 3 above)
Comment (ajgokhale) [reply python/ray/worker.py]: Completely missed that, thanks for the catch

**Comment by robertnishihara** [comment]
FYI @AdamGleave this should fix #4029.

File: python/ray/worker.py [hunk 6]
```diff
@@ -949,6 +949,10 @@ def _wait_for_and_process_task(self, task):
             task.function_descriptor_list())
         driver_id = task.driver_id()
 
+        # Check if the active driver has changed
+        previous_driver_id = self.task_driver_id
```
Comment (robertnishihara) [inline python/ray/worker.py]: I don't think this correctly identifies the previous driver ID. I think the Worker class may actually need to store the previous driver ID.

I tried out the PR but it still prints excessive amounts of `Ray driver id: DriverID(7af9149c78f6e8b93cd0c5b49af962cd7b7e9d21)` even when there is only one driver.

File: python/ray/worker.py (hunk 3 above)
Comment (robertnishihara) [reply python/ray/worker.py]: I just tried the latest version, and this is still getting printed once per task for me.

**Comment by robertnishihara** [comment]
@ajgokhale it looks like the test you modified is failing in Travis.

File: python/ray/tests/test_multi_node.py [hunk 7]
```diff
@@ -201,6 +202,70 @@ def f():
...
+            sys.exit()
+    else:
+        if str(i) in output:
+            sys.exit()
+
+error_lines = captured["err"]
+if len(error_lines) != 0:
+    sys.exit()
+
+print("success")
```
Comment (robertnishihara) [inline python/ray/tests/test_multi_node.py]: We aren't checking for this anywhere are we? If not, then remove it.

File: python/ray/tests/utils.py [hunk 8]
```diff
@@ -94,3 +94,57 @@ def wait_for_errors(error_type, num_errors, timeout=10):
             return
         time.sleep(0.1)
     raise Exception("Timing out of wait.")
+
+
+def wait_for_success_output(process_handle, timeout=10):
```
Comment (robertnishihara) [inline python/ray/tests/utils.py]: This is only used in the multi_node_test, right? Let's move it back into that file.

File: python/ray/log_monitor.py [hunk 9]
```diff
@@ -170,15 +172,38 @@ def check_log_files_and_publish_updates(self):
             assert not file_info.file_handle.closed
 
             lines_to_publish = []
+            driver_switches = []
+            if file_info.current_driver_id is not None:
+                driver_switches.append((-1, file_info.current_driver_id))
```
[the remark this answers is not in the document]
Comment (ajgokhale) [reply python/ray/log_monitor.py]: This represents the index of the line where the driver switch message is printed, which means that the actual log messages begin on the next line i.e. we add 1 to this value to find the first actual log message.

File: python/ray/tests/test_multi_node.py [hunk 10]
```diff
@@ -201,6 +202,70 @@ def f():
...
+"""
```
[the remark this answers is not in the document]
Comment (ajgokhale) [reply python/ray/tests/test_multi_node.py]: In this case, we'd like to run this script with a different range of values each time, so we format it differently during each iteration of the loop.

File: python/ray/log_monitor.py [hunk 11]
```diff
@@ -190,20 +215,28 @@ def check_log_files_and_publish_updates(self):
                         lines_to_publish[0].startswith("Ray worker pid: ")):
                     file_info.worker_pid = int(
                         lines_to_publish[0].split(" ")[-1])
-                    lines_to_publish = lines_to_publish[1:]
+                    if driver_switches[0][0] == -1:
```
[the remark this answers is not in the document]
Comment (ajgokhale) [reply python/ray/log_monitor.py]: Great point, it's definitely unnecessary

File: python/ray/tests/test_multi_node.py (hunk 7 above)
Comment (ajgokhale) [reply python/ray/tests/test_multi_node.py]: This is actually checked for in `wait_for_success_output`. I think `test_drivers_release_resources` uses a similar check to determine if the test passed.

File: python/ray/tests/test_multi_node.py [hunk 12]
```diff
@@ -201,6 +202,70 @@ def f():
...
+
+    driver_script = """
+from __future__ import print_function
+import ray
+from ray.tests.utils import CaptureOutputAndError
+import sys
+import time
+import atexit
+
+atexit.register(lambda: print("error"))
```
[the remark this answers is not in the document]
Comment (ajgokhale) [reply python/ray/tests/test_multi_node.py]: This is checked for in `wait_for_success_output` (it's a modification of the original function implementation). I realized that the original version has to timeout waiting for "success" to be printed, which, in the case of a failing test, unnecessarily wastes time.

File: python/ray/tests/test_multi_node.py [hunk 13]
```diff
@@ -201,6 +202,70 @@ def f():
...
+
+captured = dict()
+with CaptureOutputAndError(captured):
+    ray.get(f.remote())
+    time.sleep(1)
+
+output_lines = captured["out"]
+output = [line.split(" ")[1] for line in output_lines.splitlines()]
+
+for i in range(*total_range):
```
[the remark this answers is not in the document]
Comment (ajgokhale) [reply python/ray/tests/test_multi_node.py]: Fixed

File: python/ray/tests/test_multi_node.py [hunk 14]
```diff
@@ -201,6 +201,69 @@ def f():
         assert out.count(log_message) == 4
 
 
+def wait_for_success_output(process_handle, timeout=10):
```
Comment (robertnishihara) [inline python/ray/tests/test_multi_node.py]: This appears unused. it is redefined below, right?

File: python/ray/log_monitor.py [hunk 15]
```diff
@@ -170,15 +172,38 @@ def check_log_files_and_publish_updates(self):
             assert not file_info.file_handle.closed
 
             lines_to_publish = []
+            driver_switches = []
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply python/ray/log_monitor.py]: Please make this even clearer. It's not obvious to me from this comment that we need to add `1` to `index` to find such and such. The comment should be completely explicit so that people understand exactly how to use and manipulate the values in `driver_switches` without having to read the code.

File: python/ray/tests/test_multi_node.py [hunk 16]
```diff
@@ -201,6 +201,69 @@ def f():
...
+
+
+def test_logging_to_multiple_drivers(call_ray_start):
+    redis_address = call_ray_start
+
+    driver_script = """
+from __future__ import print_function
+import ray
+from ray.tests.utils import CaptureOutputAndError
+import sys
```
Comment (robertnishihara) [inline python/ray/tests/test_multi_node.py]: no longer used

File: python/ray/worker.py [hunk 17]
```diff
@@ -388,17 +391,17 @@ def put_object(self, object_id, value):
...
             # may not be serializable for cloudpickle. So we need these extra
             # fallbacks here to start from the beginning. Hopefully the object
             # could have a `__reduce__` method.
             register_custom_serializer(type(value), use_pickle=True)
-            warning_message = ("WARNING: Serializing the class {} failed, "
-                               "so are are falling back to cloudpickle."
-                               .format(type(value)))
+            warning_message = (
+                "WARNING: Serializing the class {} failed, "
+                "so are are falling back to cloudpickle.".format(type(value)))
```
Comment (robertnishihara) [inline python/ray/worker.py]: revert this change

**Comment by richardliaw** [comment]
@robertnishihara this is often brought up as a bug - any plan on merging?

**Review by richardliaw — CHANGES_REQUESTED** [review]
This needs to be rebased to convert all driver -> job
